File: src/responder/authentication.cpp
#include "opaque/responder.h"
#include "opaque/protocol.h"
#include "opaque/pq.h"
#include "opaque/secure_cleanup.h"
#include "opaque/debug_log.h"
#include <sodium.h>
#include <algorithm>

namespace ecliptix::security::opaque::responder {

KE2::KE2() : responder_nonce(NONCE_LENGTH),
             responder_public_key(PUBLIC_KEY_LENGTH),
             credential_response(CREDENTIAL_RESPONSE_LENGTH),
             responder_mac(MAC_LENGTH),
             kem_ciphertext(pq_constants::KEM_CIPHERTEXT_LENGTH) {}


ResponderState::ResponderState()
    : responder_private_key(PRIVATE_KEY_LENGTH)
    , responder_public_key(PUBLIC_KEY_LENGTH)
    , responder_ephemeral_private_key(PRIVATE_KEY_LENGTH)
    , responder_ephemeral_public_key(PUBLIC_KEY_LENGTH)
    , initiator_public_key(PUBLIC_KEY_LENGTH)
    , session_key(0)
    , expected_initiator_mac(MAC_LENGTH)
    , master_key(0)
    , handshake_complete(false)
    , pq_shared_secret(0) {}


ResponderState::~ResponderState() {
    secure_wipe(responder_private_key);
    secure_wipe(responder_public_key);
    secure_wipe(responder_ephemeral_private_key);
    secure_wipe(responder_ephemeral_public_key);
    secure_wipe(initiator_public_key);
    secure_clear(session_key);
    secure_wipe(expected_initiator_mac);
    secure_clear(master_key);
    secure_clear(pq_shared_secret);
}


Result responder_finish_impl(const uint8_t *ke3_data, size_t ke3_length,
                             ResponderState &state, secure_bytes &session_key,
                             secure_bytes &master_key) {
    log::section("RELAY: Finish (Verify KE3)");

    if (!ke3_data || ke3_length != KE3_LENGTH) {
        return Result::InvalidInput;
    }
    if (!crypto::init()) {
        return Result::CryptoError;
    }
    if (state.session_key.empty() || state.master_key.size() != MASTER_KEY_LENGTH) {
        return Result::ValidationError;
    }

    const uint8_t *initiator_mac = ke3_data;
    log::hex("ke3 initiator_mac (received)", initiator_mac, MAC_LENGTH);
    log::hex("expected_initiator_mac", state.expected_initiator_mac);

    if (crypto_verify_64(initiator_mac, state.expected_initiator_mac.data()) != 0) {
        log::msg("ERROR: Initiator MAC verification failed!");
        secure_clear(state.session_key);
        secure_clear(state.master_key);
        secure_clear(state.pq_shared_secret);
        secure_wipe(state.expected_initiator_mac);
        state.handshake_complete = false;
        return Result::AuthenticationError;
    }

    log::msg("Initiator MAC verified successfully");

    session_key = state.session_key;
    master_key = state.master_key;

    log::hex("FINAL session_key", session_key);
    log::hex("FINAL master_key", master_key);

    secure_clear(state.session_key);
    secure_clear(state.master_key);
    secure_clear(state.pq_shared_secret);
    secure_wipe(state.expected_initiator_mac);
    state.handshake_complete = true;

    return Result::Success;
}


Result generate_ke2_impl(const uint8_t *ke1_data, size_t ke1_length,
                         const ResponderCredentials &credentials,
                         const secure_bytes &responder_private_key,
                         const secure_bytes &responder_public_key,
                         const uint8_t *account_id,
                         size_t account_id_length,
                         KE2 &ke2, ResponderState &state) {
    log::section("RELAY: Generate KE2 (PQ)");

    if (!ke1_data || ke1_length != KE1_LENGTH)
        return Result::InvalidInput;
    if (!account_id || account_id_length == 0)
        return Result::InvalidInput;
    if (!crypto::init())
        return Result::CryptoError;
    if (credentials.initiator_public_key.size() != PUBLIC_KEY_LENGTH)
        return Result::InvalidPublicKey;
    if (credentials.envelope.size() != ENVELOPE_LENGTH)
        return Result::InvalidInput;

    log::hex("ke1_data (full)", ke1_data, ke1_length);
    log::hex("responder_private_key", responder_private_key);
    log::hex("responder_public_key", responder_public_key);

    protocol::Ke1View ke1_view{};
    OPAQUE_TRY(protocol::parse_ke1(ke1_data, ke1_length, ke1_view));

    const uint8_t *cred_req = ke1_view.credential_request;
    const uint8_t *init_eph_pk = ke1_view.initiator_public_key;
    const uint8_t *init_nonce = ke1_view.initiator_nonce;
    const uint8_t *init_pq_pk = ke1_view.pq_ephemeral_public_key;
    const uint8_t *init_static_pk = credentials.initiator_public_key.data();

    OPAQUE_TRY(crypto::validate_ristretto_point(cred_req, REGISTRATION_REQUEST_LENGTH));
    OPAQUE_TRY(crypto::validate_public_key(init_eph_pk, PUBLIC_KEY_LENGTH));
    OPAQUE_TRY(crypto::validate_public_key(init_static_pk, PUBLIC_KEY_LENGTH));

    std::copy_n(init_eph_pk, PUBLIC_KEY_LENGTH, state.initiator_public_key.begin());
    std::copy_n(responder_private_key.begin(), responder_private_key.size(),
                state.responder_private_key.begin());
    std::copy_n(responder_public_key.begin(), responder_public_key.size(),
                state.responder_public_key.begin());
    state.handshake_complete = false;

    SecureLocal<crypto_core_ristretto255_BYTES> evaluated_elem;
    SecureLocal<PRIVATE_KEY_LENGTH>             oprf_key;
    SecureLocal<PUBLIC_KEY_LENGTH>              dh1, dh2, dh3;
    SecureLocal<pq_constants::KEM_SHARED_SECRET_LENGTH> kem_ss;
    SecureLocal<crypto_auth_hmacsha512_BYTES>   prk;
    SecureLocal<crypto_auth_hmacsha512_BYTES>   init_mac_key;
    SecureLocal<crypto_hash_sha512_BYTES>       transcript_hash;

    auto failure_guard = make_cleanup([&] {
        secure_clear(state.session_key);
        secure_wipe(state.expected_initiator_mac);
        secure_clear(state.master_key);
        secure_clear(state.pq_shared_secret);
        state.handshake_complete = false;
    });

    do {
        crypto_core_ristretto255_scalar_random(state.responder_ephemeral_private_key.data());
    } while (sodium_is_zero(state.responder_ephemeral_private_key.data(),
                            state.responder_ephemeral_private_key.size()) == 1);

    if (crypto_scalarmult_ristretto255_base(state.responder_ephemeral_public_key.data(),
                                            state.responder_ephemeral_private_key.data()) != 0) {
        return Result::CryptoError;
    }

    OPAQUE_TRY(crypto::random_bytes(ke2.responder_nonce.data(), NONCE_LENGTH));

    std::ranges::copy(state.responder_ephemeral_public_key, ke2.responder_public_key.begin());

    OPAQUE_TRY(crypto::derive_oprf_key(responder_private_key.data(), responder_private_key.size(),
                                        account_id, account_id_length, oprf_key));

    OPAQUE_TRY(oblivious_prf::evaluate(cred_req, oprf_key, evaluated_elem));

    log::hex("evaluated_element (OPRF)", evaluated_elem.data(), evaluated_elem.size());

    {
        size_t off = 0;
        std::copy_n(evaluated_elem.data(), crypto_core_ristretto255_BYTES,
                    ke2.credential_response.begin() + static_cast<std::ptrdiff_t>(off));
        off += crypto_core_ristretto255_BYTES;
        std::ranges::copy(credentials.envelope,
                          ke2.credential_response.begin() + static_cast<std::ptrdiff_t>(off));
    }

    if (crypto_scalarmult_ristretto255(dh1, responder_private_key.data(), init_static_pk) != 0)
        return Result::CryptoError;

    if (crypto_scalarmult_ristretto255(dh2, responder_private_key.data(), init_eph_pk) != 0)
        return Result::CryptoError;

    if (crypto_scalarmult_ristretto255(dh3, state.responder_ephemeral_private_key.data(), init_static_pk) != 0)
        return Result::CryptoError;

    log::hex("dh1 (resp_priv * init_static)", dh1.data(), PUBLIC_KEY_LENGTH);
    log::hex("dh2 (resp_priv * init_eph)", dh2.data(), PUBLIC_KEY_LENGTH);
    log::hex("dh3 (resp_eph * init_static)", dh3.data(), PUBLIC_KEY_LENGTH);

    OPAQUE_TRY(pq::kem::encapsulate(init_pq_pk, ke2.kem_ciphertext.data(), kem_ss));

    state.pq_shared_secret.resize(pq_constants::KEM_SHARED_SECRET_LENGTH);
    std::copy_n(kem_ss.data(), pq_constants::KEM_SHARED_SECRET_LENGTH, state.pq_shared_secret.begin());

    log::hex("kem_shared_secret (encapsulated)", kem_ss.data(), kem_ss.size());

    {
        SecureLocal<crypto_auth_hmacsha512_BYTES> resp_mac_key;
        constexpr size_t CLASSICAL_IKM_LENGTH = 3 * PUBLIC_KEY_LENGTH;
        SecureLocal<CLASSICAL_IKM_LENGTH> classical_ikm;
        std::copy_n(dh1.data(), PUBLIC_KEY_LENGTH, classical_ikm.data());
        std::copy_n(dh2.data(), PUBLIC_KEY_LENGTH, classical_ikm.data() + PUBLIC_KEY_LENGTH);
        std::copy_n(dh3.data(), PUBLIC_KEY_LENGTH, classical_ikm.data() + 2 * PUBLIC_KEY_LENGTH);

        constexpr size_t mac_input_size =
            2 * NONCE_LENGTH + 4 * PUBLIC_KEY_LENGTH + CREDENTIAL_RESPONSE_LENGTH
            + pq_constants::KEM_CIPHERTEXT_LENGTH + pq_constants::KEM_PUBLIC_KEY_LENGTH;

        secure_bytes mac_input(mac_input_size);
        size_t off = 0;

        auto append = [&](const uint8_t* src, size_t len) {
            std::copy_n(src, len, mac_input.begin() + static_cast<std::ptrdiff_t>(off));
            off += len;
        };

        append(init_eph_pk, PUBLIC_KEY_LENGTH);
        append(state.responder_ephemeral_public_key.data(), PUBLIC_KEY_LENGTH);
        append(init_nonce, NONCE_LENGTH);
        append(ke2.responder_nonce.data(), NONCE_LENGTH);
        append(init_static_pk, PUBLIC_KEY_LENGTH);
        append(responder_public_key.data(), PUBLIC_KEY_LENGTH);
        append(ke2.credential_response.data(), CREDENTIAL_RESPONSE_LENGTH);
        append(init_pq_pk, pq_constants::KEM_PUBLIC_KEY_LENGTH);
        append(ke2.kem_ciphertext.data(), pq_constants::KEM_CIPHERTEXT_LENGTH);

        {
            crypto_hash_sha512_state ts;
            crypto_hash_sha512_init(&ts);
            crypto_hash_sha512_update(&ts,
                reinterpret_cast<const uint8_t*>(labels::kTranscriptContext),
                labels::kTranscriptContextLength);
            crypto_hash_sha512_update(&ts, mac_input.data(), mac_input.size());
            crypto_hash_sha512_final(&ts, transcript_hash);
        }

        OPAQUE_TRY(pq::combine_key_material(
            classical_ikm.data(), classical_ikm.size(),
            kem_ss, kem_ss.size(),
            transcript_hash, transcript_hash.size(),
            prk));

        log::hex("prk (PQ combiner)", prk.data(), prk.size());

        state.session_key.resize(HASH_LENGTH);
        OPAQUE_TRY(crypto::key_derivation_expand(prk, prk.size(),
            reinterpret_cast<const uint8_t*>(pq::labels::kPqSessionKeyInfo),
            pq::labels::kPqSessionKeyInfoLength,
            state.session_key.data(), state.session_key.size()));

        state.master_key.resize(MASTER_KEY_LENGTH);
        OPAQUE_TRY(crypto::key_derivation_expand(prk, prk.size(),
            reinterpret_cast<const uint8_t*>(pq::labels::kPqMasterKeyInfo),
            pq::labels::kPqMasterKeyInfoLength,
            state.master_key.data(), state.master_key.size()));

        OPAQUE_TRY(crypto::key_derivation_expand(prk, prk.size(),
            reinterpret_cast<const uint8_t*>(pq::labels::kPqResponderMacInfo),
            pq::labels::kPqResponderMacInfoLength,
            resp_mac_key, resp_mac_key.size()));

        OPAQUE_TRY(crypto::hmac(resp_mac_key, resp_mac_key.size(),
                                 mac_input.data(), mac_input.size(),
                                 ke2.responder_mac.data()));

        log::hex("ke2.responder_mac", ke2.responder_mac);

        OPAQUE_TRY(crypto::key_derivation_expand(prk, prk.size(),
            reinterpret_cast<const uint8_t*>(pq::labels::kPqInitiatorMacInfo),
            pq::labels::kPqInitiatorMacInfoLength,
            init_mac_key, init_mac_key.size()));

        OPAQUE_TRY(crypto::hmac(init_mac_key, init_mac_key.size(),
                                 mac_input.data(), mac_input.size(),
                                 state.expected_initiator_mac.data()));

        log::hex("expected_initiator_mac", state.expected_initiator_mac);

        sodium_memzero(classical_ikm.data(), classical_ikm.size());
        sodium_memzero(mac_input.data(), mac_input.size());
    }

    failure_guard.dismiss();

    return Result::Success;
}

} // namespace ecliptix::security::opaque::responder

File: rust/crates/opaque-relay/src/authentication.rs
// Copyright (c) 2026 Oleksandr Melnychenko, Ukraine
// Ecliptix Security — Hybrid PQ-OPAQUE
// Licensed under the MIT License

use opaque_core::types::{
    constant_time_eq, labels, pq, pq_labels, CREDENTIAL_RESPONSE_LENGTH, ENVELOPE_LENGTH,
    HASH_LENGTH, KE1_LENGTH, KE3_LENGTH, MAC_LENGTH, MASTER_KEY_LENGTH, NONCE_LENGTH, OpaqueError,
    OpaqueResult, PRIVATE_KEY_LENGTH, PUBLIC_KEY_LENGTH,
};
use opaque_core::{crypto, oprf, pq_kem, protocol};
use zeroize::Zeroize;

use crate::state::{Ke2Message, OpaqueResponder, ResponderCredentials, ResponderState};

pub fn generate_ke2(
    responder: &OpaqueResponder,
    ke1_data: &[u8],
    account_id: &[u8],
    credentials: &ResponderCredentials,
    ke2: &mut Ke2Message,
    state: &mut ResponderState,
) -> OpaqueResult<()> {
    if ke1_data.len() != KE1_LENGTH {
        return Err(OpaqueError::InvalidProtocolMessage);
    }
    if account_id.is_empty() {
        return Err(OpaqueError::InvalidInput);
    }
    if credentials.initiator_public_key.iter().all(|&b| b == 0) {
        return Err(OpaqueError::InvalidPublicKey);
    }
    if credentials.envelope.len() != ENVELOPE_LENGTH {
        return Err(OpaqueError::InvalidEnvelope);
    }

    let ke1 = protocol::parse_ke1(ke1_data)?;
    let kp = responder.keypair();

    let init_eph_pk: &[u8; PUBLIC_KEY_LENGTH] = ke1.initiator_public_key
        .try_into().map_err(|_| OpaqueError::InvalidProtocolMessage)?;
    let init_static_pk = &credentials.initiator_public_key;

    crypto::validate_ristretto_point(ke1.credential_request)?;
    crypto::validate_public_key(init_eph_pk)?;
    crypto::validate_public_key(init_static_pk)?;

    state.initiator_public_key = *init_eph_pk;
    state.responder_private_key = kp.private_key;
    state.responder_public_key = kp.public_key;
    state.handshake_complete = false;

    state.responder_ephemeral_private_key = crypto::random_nonzero_scalar();
    state.responder_ephemeral_public_key = crypto::scalarmult_base(&state.responder_ephemeral_private_key)?;

    crypto::random_bytes(&mut ke2.responder_nonce)?;
    ke2.responder_public_key = state.responder_ephemeral_public_key;

    let mut oprf_key = [0u8; PRIVATE_KEY_LENGTH];
    crypto::derive_oprf_key(responder.oprf_seed(), account_id, &mut oprf_key)?;

    let cred_req: &[u8; PUBLIC_KEY_LENGTH] = ke1.credential_request
        .try_into().map_err(|_| OpaqueError::InvalidProtocolMessage)?;
    let mut evaluated_elem = [0u8; PUBLIC_KEY_LENGTH];
    oprf::evaluate(cred_req, &oprf_key, &mut evaluated_elem)?;
    oprf_key.zeroize();

    ke2.credential_response[..PUBLIC_KEY_LENGTH].copy_from_slice(&evaluated_elem);
    ke2.credential_response[PUBLIC_KEY_LENGTH..].copy_from_slice(&credentials.envelope);

    let mut dh1 = [0u8; PUBLIC_KEY_LENGTH];
    let mut dh2 = [0u8; PUBLIC_KEY_LENGTH];
    let mut dh3 = [0u8; PUBLIC_KEY_LENGTH];
    // 4th DH: resp_eph × init_eph — provides unknown key share (UKS) resistance
    let mut dh4 = [0u8; PUBLIC_KEY_LENGTH];

    crypto::scalar_mult(&kp.private_key, init_static_pk, &mut dh1)?;
    crypto::scalar_mult(&kp.private_key, init_eph_pk, &mut dh2)?;
    crypto::scalar_mult(&state.responder_ephemeral_private_key, init_static_pk, &mut dh3)?;
    crypto::scalar_mult(&state.responder_ephemeral_private_key, init_eph_pk, &mut dh4)?;

    let mut kem_ss = [0u8; pq::KEM_SHARED_SECRET_LENGTH];
    pq_kem::encapsulate(ke1.pq_ephemeral_public_key, &mut ke2.kem_ciphertext, &mut kem_ss)?;
    state.pq_shared_secret = kem_ss.to_vec();

    let mut classical_ikm = [0u8; 4 * PUBLIC_KEY_LENGTH];
    classical_ikm[..PUBLIC_KEY_LENGTH].copy_from_slice(&dh1);
    classical_ikm[PUBLIC_KEY_LENGTH..2 * PUBLIC_KEY_LENGTH].copy_from_slice(&dh2);
    classical_ikm[2 * PUBLIC_KEY_LENGTH..3 * PUBLIC_KEY_LENGTH].copy_from_slice(&dh3);
    classical_ikm[3 * PUBLIC_KEY_LENGTH..].copy_from_slice(&dh4);

    let mac_input_size = 2 * NONCE_LENGTH
        + 4 * PUBLIC_KEY_LENGTH
        + CREDENTIAL_RESPONSE_LENGTH
        + pq::KEM_CIPHERTEXT_LENGTH
        + pq::KEM_PUBLIC_KEY_LENGTH;
    let mut mac_input = vec![0u8; mac_input_size];
    let mut off = 0;

    let mut append = |data: &[u8]| {
        mac_input[off..off + data.len()].copy_from_slice(data);
        off += data.len();
    };
    append(init_eph_pk);
    append(&state.responder_ephemeral_public_key);
    append(ke1.initiator_nonce);
    append(&ke2.responder_nonce);
    append(init_static_pk);
    append(&kp.public_key);
    append(&ke2.credential_response);
    append(ke1.pq_ephemeral_public_key);
    append(&ke2.kem_ciphertext);

    let mut transcript_hash = [0u8; HASH_LENGTH];
    crypto::sha512_multi(&[labels::TRANSCRIPT_CONTEXT, &mac_input], &mut transcript_hash);

    let mut prk = [0u8; HASH_LENGTH];
    pq_kem::combine_key_material(&classical_ikm, &kem_ss, &transcript_hash, &mut prk)?;

    state.session_key = vec![0u8; HASH_LENGTH];
    crypto::key_derivation_expand(&prk, pq_labels::PQ_SESSION_KEY_INFO, &mut state.session_key)?;

    state.master_key = vec![0u8; MASTER_KEY_LENGTH];
    crypto::key_derivation_expand(&prk, pq_labels::PQ_MASTER_KEY_INFO, &mut state.master_key)?;

    let mut resp_mac_key = [0u8; MAC_LENGTH];
    crypto::key_derivation_expand(&prk, pq_labels::PQ_RESPONDER_MAC_INFO, &mut resp_mac_key)?;
    crypto::hmac_sha512(&resp_mac_key, &mac_input, &mut ke2.responder_mac)?;

    let mut init_mac_key = [0u8; MAC_LENGTH];
    crypto::key_derivation_expand(&prk, pq_labels::PQ_INITIATOR_MAC_INFO, &mut init_mac_key)?;
    crypto::hmac_sha512(&init_mac_key, &mac_input, &mut state.expected_initiator_mac)?;

    classical_ikm.zeroize();
    mac_input.zeroize();
    prk.zeroize();
    dh1.zeroize();
    dh2.zeroize();
    dh3.zeroize();
    dh4.zeroize();
    kem_ss.zeroize();
    resp_mac_key.zeroize();
    init_mac_key.zeroize();

    Ok(())
}

pub fn responder_finish(
    ke3_data: &[u8],
    state: &mut ResponderState,
    session_key: &mut Vec<u8>,
    master_key: &mut Vec<u8>,
) -> OpaqueResult<()> {
    if ke3_data.len() != KE3_LENGTH {
        return Err(OpaqueError::InvalidProtocolMessage);
    }
    if state.session_key.is_empty() || state.master_key.len() != MASTER_KEY_LENGTH {
        return Err(OpaqueError::ValidationError);
    }

    let initiator_mac = ke3_data;

    if !constant_time_eq(initiator_mac, &state.expected_initiator_mac) {
        state.session_key.zeroize();
        state.master_key.zeroize();
        state.pq_shared_secret.zeroize();
        state.expected_initiator_mac.zeroize();
        state.handshake_complete = false;
        return Err(OpaqueError::AuthenticationError);
    }

    *session_key = std::mem::take(&mut state.session_key);
    *master_key = std::mem::take(&mut state.master_key);

    state.pq_shared_secret.zeroize();
    state.expected_initiator_mac.zeroize();
    state.handshake_complete = true;

    Ok(())
}
